streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -205,12 +205,12 @@ boolean setState(final State newState) {
         return true;
     }
 
-    public synchronized boolean isRunningAndNotRebalancing() {
+    public boolean isRunningAndNotRebalancing() {
         // we do not need to grab stateLock since it is a single read
         return state == State.RUNNING;
     }
 
-    public synchronized boolean isRunning() {
```

Comment:
For this function, logically it reads the field `state` multiple times, while depending on the compilation it may be just read once and cached in register, but we cannot guarantee that. So I'm not sure if it is safe to remove the `synchronized` keyword.

Reply:
Let us look at this sequence for the current code:

t1: state == State.PARTITIONS_REVOKED
t2: if statement checks against State.RUNNING - false
t3: state changes to State.RUNNING
t4: if statement checks against State.PARTITIONS_REVOKED - false

The above is just example illustrating what may happen at runtime.
So the proposed form is no worse than the current.

Reply:
The read of state variable is protected by stateLock.
So there is no issue dropping the synchronized keyword.

Reply:
Makes sense. Thanks.